[MRG] Update `quantile_transform` `copy` default to True

HunterMcGushion opened the pull request:
#### What does this implement/fix? Explain your changes.
- Changes default value for the `copy` kwarg in :func:`preprocessing.data.quantile_transform` from False to True
- The original default of `copy=False` was in conflict with the function's documentation:
https://github.com/scikit-learn/scikit-learn/blob/ffc63c6508817112ebb5157fc36696d4bf00e53b/sklearn/preprocessing/data.py#L2492
- This was the only function in the module, whose default `copy` kwarg value was False
- Seems like True was the intended default, since performing an in-place transformation by default is pretty unexpected - At least, it was for me

#### Any other comments?
- If the signature was correct, and `copy=False` was the intended default, I'd be happy to update the documentation instead
- In any case, the documentation conflicts with the actual signature

HunterMcGushion committed:
Update `quantile_transform` `copy` default to True

qinhanmin2014 reviewed (commented):
Thanks for the PR, we need a deprecation cycle here, see https://scikit-learn.org/dev/developers/contributing.html#change-the-default-value-of-a-parameter

HunterMcGushion commented:
Got it. Sorry for overlooking that. What version should the `FutureWarning` list for the change to take place? 

Would it make sense to change the docstring to at least match the current default, since the two are still in conflict?

qinhanmin2014 commented:
> What version should the FutureWarning list for the change to take place?

0.23 (next release is 0.21)

> Would it make sense to change the docstring to at least match the current default, since the two are still in conflict?

We usually add some notes in the docstring, e.g., see https://scikit-learn.org/dev/modules/generated/sklearn.metrics.auc.html.

HunterMcGushion commented:
Alright, so leave line 2492 unchanged, and just add a note that the default value stated there is incorrect and will be fixed in 0.23?

qinhanmin2014 commented:
> Alright, so leave line 2492 unchanged, and just add a note that the default value stated there is incorrect and will be fixed in 0.23?

Please refer to  https://scikit-learn.org/dev/developers/contributing.html#change-the-default-value-of-a-parameter

HunterMcGushion commented:
Ah, of course. I feel quite ridiculous. Thanks for pointing me in the right direction again

jnothman commented:
Thanks for raising this! I agree it's an unnecessary surprise.

HunterMcGushion committed:
Add FutureWarning for `quantile_transform` default `copy`

HunterMcGushion committed:
Address flake8 complaint about line being too long

HunterMcGushion commented:
Thanks, @qinhanmin2014 and @jnothman! How does this look?

qinhanmin2014 commented:
> So just to make sure I'm following the guidelines, I should just wrap the failing tests with @pytest.mark.filterwarnings(...), correct?

Yes, please. Maybe add some comments, e.g., ``@pytest.mark.filterwarnings(...)  # 0.23``

The deprecation message seems strange, please try to follow the contribution guide. (e.g., ``The default value of XXX will change from XXX to XXX in XXX in order to.../because...``)

HunterMcGushion committed:
Make deprecation message follow contributing guidelines

HunterMcGushion committed:
Use explicit `copy=False` in docstring example

- This change was made to prevent `FutureWarning` from being raised
by the example

HunterMcGushion committed:
Filter warnings pertaining to updated default `copy` value

HunterMcGushion commented:
@qinhanmin2014, regarding 08faa3ce3c8d4f561b4fa5e4984b03aebcd5ee02, is this the proper way to deal with deprecations in the docstring examples? Or should I use `warnings.filterwarnings`?

qinhanmin2014 commented on sklearn/preprocessing/data.py:
```diff
@@ -2489,9 +2489,19 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+            make it more consistent with the default `copy` values of other
+            functions in :mod:`sklearn.preprocessing.data`. Furthermore, the
+            current default of False may have unexpected side effects by
+            modifying the value of `X` inplace. For more information see
```
you don't need to reference the PR.

qinhanmin2014 commented on doc/whats_new/v0.21.rst:
```diff
@@ -458,6 +458,13 @@ Support for Python 3.4 and below has been officially dropped.
   useless or resulting in a wrong approximation of the cumulative distribution
   function estimator. :issue:`13333` by :user:`Albert Thomas <albertcthomas>`.
 
+- |API| Deprecate default `copy=False` value of
```
this seems strange, maybe copy the deprecation message here.
``preprocessing.data.quantile_transform`` -> ``preprocessing.quantile_transform``

qinhanmin2014 commented on sklearn/preprocessing/data.py:
```diff
@@ -2507,7 +2517,7 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
     >>> rng = np.random.RandomState(0)
     >>> X = np.sort(rng.normal(loc=0.5, scale=0.25, size=(25, 1)), axis=0)
-    >>> quantile_transform(X, n_quantiles=10, random_state=0)
+    >>> quantile_transform(X, n_quantiles=10, random_state=0, copy=False)
```
maybe set copy=True here

HunterMcGushion committed:
Use `copy=True` in example and remove PR reference

HunterMcGushion committed:
Reformat whats_new entry to match deprecation message

qinhanmin2014 reviewed (approved):
LGTM, thanks @HunterMcGushion

jnothman commented:
Thanks @HunterMcGushion!
